substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -127,3 +140,738 @@ where
+				continue;
+			};
+
+			let maybe_result = match operation_type {
+				ArchiveStorageDiffOperationType::Added =>
+					self.fetch_storage(hash, key.clone(), maybe_child_trie.clone(), fetch_type)?,
+				ArchiveStorageDiffOperationType::Deleted => self.fetch_storage(
+					previous_hash,
+					key.clone(),
+					maybe_child_trie.clone(),
+					fetch_type,
+				)?,
+				ArchiveStorageDiffOperationType::Modified => {
+					let Some(storage_result) = self.fetch_storage(
+						hash,
+						key.clone(),
+						maybe_child_trie.clone(),
+						fetch_type,
+					)?
+					else {
+						continue
+					};
+
+					let Some(previous_storage_result) = self.fetch_storage(
+						previous_hash,
+						key.clone(),
+						maybe_child_trie.clone(),
+						fetch_type,
+					)?
+					else {
+						continue
+					};
+
+					// For modified records we need to check the actual storage values.
+					if storage_result == previous_storage_result {
+						continue
+					}
+
+					Some(storage_result)
+				},
+			};
+
+			if let Some(storage_result) = maybe_result {
+				if !Self::send_result(
+					&tx,
+					storage_result,
+					operation_type,
+					maybe_child_trie_str.clone(),
+				) {
+					return Ok(())
+				}
+			}
+		}
+
+		Ok(())
+	}
+
+	/// The items provided to this method are obtained by calling `deduplicate_storage_diff_items`.
+	/// The deduplication method ensures that all items `Vec<DiffDetails>` correspond to the same
+	/// `child_trie_key`.
+	///
+	/// This method will iterate over the keys of the main trie or a child trie and fetch the
+	/// given keys. The fetched keys will be sent to the provided `tx` sender to leverage
+	/// the backpressure mechanism.
+	pub async fn handle_trie_queries(
+		&self,
+		hash: Block::Hash,
+		previous_hash: Block::Hash,
+		trie_queries: Vec<Vec<DiffDetails>>,
+		tx: mpsc::Sender<ArchiveStorageDiffEvent>,
+	) -> Result<(), tokio::task::JoinError> {
+		let this = ArchiveStorageDiff { client: self.client.clone() };
+
+		tokio::task::spawn_blocking(move || {
+			for items in trie_queries {
+				log::trace!(
+					target: LOG_TARGET,
+					"handle_trie_queries: hash={:?}, previous_hash={:?}, items={:?}",
+					hash,
+					previous_hash,
+					items
+				);
+
+				let result = this.handle_trie_queries_inner(hash, previous_hash, items, &tx);
+
+				if let Err(error) = result {
+					log::trace!(
+						target: LOG_TARGET,
+						"handle_trie_queries: sending error={:?}",
+						error,
+					);
+
+					let _ = tx.blocking_send(ArchiveStorageDiffEvent::err(error));
+
+					return
+				} else {
+					log::trace!(
+						target: LOG_TARGET,
+						"handle_trie_queries: sending storage diff done",
+					);
+				}
+			}
+
+			let _ = tx.blocking_send(ArchiveStorageDiffEvent::StorageDiffDone);
+		})
+		.await?;
+
+		Ok(())
+	}
+}
+
+/// Deduplicate the provided items and return a list of `DiffDetails`.
+///
+/// Each list corresponds to a single child trie or the main trie.
+pub fn deduplicate_storage_diff_items(
```

Comment:
If this happened at the beggining of handle_trie_queries it wouldn't need to be pub I guess; I wonder whether that would complicate returning the error or whatever though.

Reply:
Yep that makes sense, have moved this on the spawn_blocking instead 🙏